### PR TITLE
feat(nat): add native video memory service

### PR DESCRIPTION
**Pull request by nvddr**
## Summary
- add typed `video-memory-service` ownership for live hub frames and recorded-video storage/decoding
- expose the capability as native `xr_video_memory` NAT functions
- reduce `video-mcp` to an optional compatibility adapter while preserving its conditional legacy tool sets
- start the typed service before the adapter in XR render demo

## Why
A process boundary should not force NAT-native agents through MCP. This keeps raw hub IPC and NVDEC in one long-running capability service, gives NAT callers a typed native function group, and retains MCP only for non-NAT consumers.

## Validation
- `uvx ruff check ...`
- `uv run --project tests pytest -q tests/test_video_memory_functions.py tests/test_xr_tracking_functions.py tests/test_spatial_math_functions.py` (10 passed)
- package `uv lock` checks for xr-ai-nat, video-mcp, video-memory-service, and tests
- import and compile checks
- GPU end-to-end tests updated for the service + adapter lifecycle and remain hardware-gated

## Stack
Depends on #288.

### CODE REVIEW

**Review by wenxind-nvidia — COMMENTED** [review]
## Review — native video-memory service (formative, DRAFT)

Reviewed the incremental diff against `migration/03-native-xr-tracking`: the new `services/video-memory-service/` (service/store/frames/live/entry), the `xr_video_memory` NAT function group (`functions/video_memory/*`), the gutted `video-mcp` compatibility adapter, docs/`DEPENDENCIES.md`/`AGENTS.md`/changelog, and tests.

**Verdict (draft):** Solid, faithful port. The native service reproduces `video-mcp`'s contract — `get_frame_from_time` live-vs-recorded anchoring, `get_latest_frame` key set, chunk-window selection in `query`, stats math, `.identity` sidecar handling, and BT.601 YUV→RGB all carry over unchanged. Scope is coherent for one step: the `-704` in `video_mcp_server/__main__.py` is that file collapsing into a thin RPC adapter, and docs/`DEPENDENCIES.md`/`AGENTS.md`/changelog are all updated in the same commit. Lifecycle is clean (hub `ProcessorEndpoint` task cancelled in `finally`; `RPCServer.close()` drains in-flight tasks and closes the socket with `LINGER 0`; client closed in both the group `finally` and the adapter `finally`). Path-escape (`_check`) checks are preserved. SPDX headers present on every new file. No blockers.

### Blocker
none

### Suggestion

1. **Native function group loses the "hide historical tools when recording disabled" property.** `functions/video_memory/functions.py` registers all five functions unconditionally. The MCP adapter still honors the original design — when recording is off it only exposes the live tool set (test `test_video_mcp_preserves_conditional_tool_sets` confirms). But native agents always see `query_video` / `get_video_stats` / `list_recorded_participants` and only discover recording is disabled at call time via a `recording_disabled` RPCError. The original `video-mcp` docstring was explicit that historical tools are "omitted entirely so the LLM never sees them and cannot attempt to call them." Registration is static in NAT, so exact parity is hard — but `video_memory_functions` already needs a live `client`, so it could `await client.get_health()` at group setup and conditionally `add_function` the recorded ops (mirroring the adapter, which already reads `recording_enabled`). At minimum, document this intended divergence.

2. **No CPU test covers `get_frame_from_time`.** The CPU test exercises `store`, the adapter tool-set gating, and native `list_*`/`get_video_stats`/`query_video`, which is good coverage for a draft. But neither the live path nor the recorded path of `get_frame_from_time` is tested off-GPU — not even the graceful error paths (`not_found` when `fetch_latest` returns `None`, `recording_disabled` when `store is None`). `_NoLiveFrames.fetch_latest` already returns `None`, so a CPU test asserting the live-frame RPCError surfaces cleanly through both the native function and the adapter would be cheap and would lock in the error contract.

### Nit

1. Inconsistent error-dict construction in the adapter (`video_mcp_server/__main__.py`): `get_latest_frame` and `get_video_stats` use the `_error()` helper, while `get_frame_from_time` and `query_video` inline `{"error": str(error)}` (because they also catch `ValueError` for request validation, and `_error` is typed to `RPCError`). Widening `_error` to accept `Exception` would let all four read the same.

2. `out_dir` accumulates `.png` / `.264` query artifacts with no eviction. Carried over from the original `video-mcp`, so not a regression — but now that a single long-running service centralizes this write path, a TTL/cleanup policy is worth a follow-up rather than unbounded growth in `/tmp` or the run dir.

Nice touches worth keeping: `QueryVideoRequest` now validates `end_us >= start_us` (the old tool didn't); `ge=0` bounds on `second_ago` / `reference_time_us` moved into the schema; and startup ordering is correct by construction — `main.py` lists `video-memory` before `video-mcp`, the launcher gates plain `Process` entries sequentially on the ready-file, and `RPCServer` touches ready only after the socket binds, so the adapter's `get_health()` at startup won't race the service.

**Comment by wenxind-nvidia** [comment]
## Re-review vs target modularization

Checking each structural choice for convergence (CONVERGES / INTERMEDIATE-OK / DEVIATES).

**Verdict:** Solid intermediate — correctly adopts the native-function-group + typed-service pattern; conventions all met (SPDX/DCO/DEPENDENCIES/AGENTS/changelog in-commit). But it **diverges from the target in three structural choices** and the known **frame-selection index math is not CPU-tested**.

### Direction vs target

| Choice | Classification | Evidence |
|---|---|---|
| Native `register_function_group` + typed `_client` + config `xr_video_memory` | **CONVERGES** | Target `functions.py` uses the same pattern/config name. |
| Register all query tools unconditionally in the group | **CONVERGES** | Target also registers unconditionally; conditional gating stays only in the MCP surface. |
| Keep `agent-mcp-servers/video-mcp` as a thin compat adapter | **INTERMEDIATE-OK** | Target has **no** video-mcp at all — deleted downstream; retaining mid-migration is fine. |
| Separate `functions/video_memory/schemas.py` | **DEVIATES (minor)** | Target inlines the models into **`_client.py`** (not `functions.py`); `video_memory/` = `__init__` + `_client` + `functions` only. Naming differs too (PR `FrameAtTimeRequest/FrameResult/EmptyRequest/ParticipantsResult` vs target `HistoricalFrameRequest/HistoricalFrameResult/ListRecordedParticipantsRequest`). |
| Multi-file service `frames.py`/`live.py`/`store.py`/`service.py`/`__main__.py` | **DEVIATES** | Target `services/video-memory-service/` is a **single `__main__.py`** (store/decode/nv12/service inline) — matching target `openxr-service`. PR splits in the opposite direction from where the target consolidates. |
| Fold **live**-frame acquisition into video-memory (`LiveFrameProvider`, `list_live_participants`, hub `ProcessorEndpoint`) | **DEVIATES** | Target `video_memory` is **recorded-only**; live-frame acquisition lives in **`xr-ai-hub-client`**. The PR even edits `AGENTS.md` to state "Video memory calls its typed service for live and recorded frames," contradicting the target's separation. This is the most notable divergence — it points away from the target's live/recorded boundary. |

Also: RPC import is `xr_ai_nat.functions._rpc` here vs target `._service.rpc` (artifact of PR #288's split).

### Findings by severity
**Important — frame-selection index math not CPU-tested (conf ~85).** `services/video-memory-service/video_memory_service/service.py`, `VideoMemoryService._recorded_frame`: the within-chunk selection (`ratio = (target_us-start_us)/(end_us-start_us)` → `index = round(ratio*(declared_frames-1))` → interpolated `timestamp_us`) is exactly the series gap, and it sits inline in a method requiring NVDEC (`decode_h264`), so `tests/test_video_memory_functions.py` can't reach it. CPU tests cover only *chunk* selection, participants, stats, `query_video`. Extract the pure `(start_us, end_us, declared_frames, target_us) → (index, timestamp_us)` computation into a helper and add a CPU test with boundary/clamp/single-frame cases.

**Conventions — pass.** SPDX on all new files; DCO present; `DEPENDENCIES.md` in-commit (new `xr-video-memory-service`, port 8310, video-mcp re-pointed, Python `<3.13` rationale rewired); entry point `xr_ai_nat_video_memory` + `tests/pyproject.toml` editable source added.

**Tool-surface parity — no issue.** `build_mcp` keeps the conditional legacy sets (live-only: `list_live_participants`/`get_frame_from_time`/`get_latest_frame`; recorded drops `get_latest_frame`, adds `list_recorded_participants`/`get_video_stats`/`query_video`), locked by `test_video_mcp_preserves_conditional_tool_sets`.

File: services/video-memory-service/video_memory_service/live.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Track live hub frame signals and fetch pixels on demand."""
+
+from xr_ai_agent import FrameData, FrameSignal, ProcessorEndpoint
```
Comment (wenxind-nvidia) [inline services/video-memory-service/video_memory_service/live.py]: Should live frame query be separated from video memory query?

File: AGENTS.md [hunk 2]
```diff
@@ -140,6 +141,8 @@ network boundary. `VisionFunctionsConfig` adds image-question answering over an
 injected `xr-ai-models` VLM while leaving frame acquisition to its own
 capability. `XRTrackingFunctionsConfig` exposes the current user frame through
 the typed OpenXR service without routing native agents through MCP.
+`VideoMemoryFunctionsConfig` exposes live frames and recorded video through a
+typed service while keeping MCP optional.
```
[the remark this answers is not in the document]
Comment (nvddr) [reply AGENTS.md]: Addressed in b2fde34: AGENTS now describes VideoMemoryFunctionsConfig as recorded-video discovery, query, and frame extraction only.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 3]
```diff
@@ -1,760 +1,165 @@
...
     @mcp.tool()
-    def list_live_participants() -> list[str]:
-        """Return raw participant identities currently connected to the
-        hub. Drawn from the live IPC roster — these are the only pids
-        for which ``get_frame_from_time(..., second_ago=0)`` will return
-        a live frame."""
-        return sorted(provider.connected_participants())
+    async def list_live_participants() -> list[str]:
+        """Return participants whose current live camera frame is available."""
+        return (await client.list_live_participants()).participants
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: No longer applicable after ecdf52c: list_live_participants no longer calls the typed video service. It reads the adapter-owned LiveFrameSource directly.

File: tests/test_video_memory_functions.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+async def _running_server(endpoint: str, dispatch):
+    server = RPCServer(endpoint, dispatch)
+    task = asyncio.create_task(server.serve())
+    await asyncio.sleep(0.02)
+    try:
+        yield
+    finally:
+        task.cancel()
+        with contextlib.suppress(asyncio.CancelledError):
+            await task
```
[the remark this answers is not in the document]
Comment (nvddr) [reply tests/test_video_memory_functions.py:123]: This await intentionally drains the cancelled RPC server task before the test releases its IPC endpoint; it is not a no-op. Keeping the cancellation wait avoids leaking the task into the next test.

File: AGENTS.md [hunk 5]
```diff
@@ -60,8 +60,9 @@ deps/               # Gitignored downloaded binaries (e.g. LOVR AppImage)
   application's explicit native-function list to MCP-only agents, but native
   applications invoke the functions directly. XR tracking calls the typed
   OpenXR service; OXR MCP only republishes the tracking and spatial functions.
-  Text-memory owns transcript JSONL storage; transcript MCP only republishes
-  that capability.
+  Video memory calls its typed service for live and recorded frames; video MCP
+  only republishes selected video-memory functions. Text-memory owns transcript
+  JSONL storage; transcript MCP only republishes that capability.
```
[the remark this answers is not in the document]
Comment (nvddr) [reply AGENTS.md]: Addressed in b2fde34: the architecture guidance now keeps live frames with the caller hub client and the typed service with recorded history.

File: docs/xr-render-demo.md [hunk 6]
```diff
@@ -254,13 +255,15 @@ On each `TranscriptionFrame`:
 | `oxr-mcp` | 8230 | `get_head_pose`, `position_ahead`, `position_relative`, `place_user_relative`, `place_object_relative`, `place_inside_by_id`, `displace_object`, `displace_objects`, `get_health` |
 | `vec-mcp` | 8250 | `between_anchors`, `world_offset`, `along_direction`, `scale_value` |
 | `vlm-mcp` | 8240 | `ask_image` |
-| `video-mcp` | 8210 | `list_live_participants`, `list_recorded_participants`, `get_video_stats`, `query_video`, `get_latest_frame`, `get_frame_at_time` |
+| `video-mcp` | 8210 | `list_live_participants`, `get_frame_from_time` (always); `list_recorded_participants`, `get_video_stats`, `query_video` (recording enabled only); `get_latest_frame` (deprecated) |
```
[the remark this answers is not in the document]
Comment (nvddr) [reply docs/xr-render-demo.md]: Addressed in b2fde34: the tool table now marks get_latest_frame as recording-disabled only and deprecated.

File: services/video-memory-service/video_memory_service/live.py (hunk 1 above)
Comment (nvddr) [reply services/video-memory-service/video_memory_service/live.py]: Addressed in ecdf52c: video-memory-service is now recorded-history only. Current live frames use the caller-owned LiveFrameSource, while Video MCP keeps the temporary live compatibility path.

File: tests/test_gpu_video_mcp.py [hunk 7]
```diff
@@ -137,6 +137,15 @@ async def _wait_ready(ready_file: pathlib.Path, proc: subprocess.Popen, timeout:
...
 
 
+def _stop_process(proc: subprocess.Popen) -> None:
+    proc.terminate()
+    try:
+        proc.wait(timeout=10)
+    except subprocess.TimeoutExpired:
+        proc.kill()
+        proc.wait(timeout=5)
+
```
[the remark this answers is not in the document]
Comment (nvddr) [reply tests/test_gpu_video_mcp.py:155]: Addressed in b2fde34: teardown now returns when the child has already exited and suppresses ProcessLookupError around terminate and kill.

File: agent-samples/xr-render-demo/yaml/xr_media_hub.yaml [hunk 8]
```diff
@@ -27,7 +27,7 @@ web_client_dir:    ../../../client-samples/web-xr
 enable_token_server: true
 
 # ── Video recording ──────────────────────────────────────────────────────────
-# Disabled: xr-render-demo uses get_latest_frame (live IPC) for visual queries.
-# Enable and set recordings_dir in video_mcp_server.yaml to unlock historical tools.
+# Disabled: xr-render-demo uses the live video-memory path for visual queries.
+# Enable this and set recordings_dir in video_memory_service.yaml for history.
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-samples/xr-render-demo/yaml/xr_media_hub.yaml]: Addressed in b2fde34: the sample comment now identifies LiveFrameSource as the current-frame path and video memory as optional recording history.

File: services/video-memory-service/video_memory_service/store.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+class ChunkStore:
+    def __init__(self, root: Path) -> None:
+        self.root = root.resolve()
+
+    def _check(self, path: Path) -> Path:
+        resolved = path.resolve()
+        if not resolved.is_relative_to(self.root):
+            raise ValueError(f"Path escapes recordings directory: {path}")
+        return resolved
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/store.py:28]: Addressed in b2fde34: path escapes now raise RPCError with the stable path_escape code, covered by a focused CPU test.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 10]
```diff
@@ -1,760 +1,165 @@
...
-              earliest_us (Unix µs), latest_us (Unix µs).
-        Returns an error dict if no chunks exist.
-        """
-        result = store.stats(participant_id)
-        if result is None:
-            return {"error": f"No video chunks for {participant_id!r}"}
-        return result
+    async def list_recorded_participants() -> list[str]:
+        """Return participants with recorded video chunks."""
+        return (await client.list_recorded_participants()).participants
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: Addressed in b2fde34: recorded discovery now logs RPCError and returns an empty list, with a focused MCP test for the fallback contract.

File: docs/changelog.md [hunk 11]
```diff
@@ -9,6 +9,23 @@ Significant decisions, in reverse-chronological order. Update this whenever a
...
+absolute Unix-microsecond reference timestamp plus a whole-second offset so an
+agent can reason coarsely while receiving the precise selected timestamp.
+
+### 2026-07-20 — Live and recorded video become native video-memory functions
+
+`xr_video_memory` exposes live-participant discovery, recorded-video queries,
+and timestamp-anchored frame extraction through `services/video-memory-service`.
+The service owns hub IPC, filesystem access, and NVDEC. Video MCP keeps port
+8210 and its conditional legacy tool list as an adapter to the typed service at
+port 8310.
```
[the remark this answers is not in the document]
Comment (nvddr) [reply docs/changelog.md]: Addressed in b2fde34: the earlier changelog entry now describes recorded history only and points to the later live-frame separation decision.

File: docs/source/components/ai-services.md [hunk 12]
```diff
@@ -27,7 +27,8 @@ tool-calling, reasoning, and hardware trade-offs documented below.
 | `ai-services/llm/nemotron3_nano/` | `nemotron3_nano_llm_server` | 8107 | NVIDIA-Nemotron-3-Nano-30B-A3B-{NVFP4,FP8} | vLLM (pip or docker) |
 | `ai-services/llm/nemotron_omni/` | `nemotron_omni_llm_server` | 8108 | Nemotron-3-Nano-Omni-30B-A3B-Reasoning (NVFP4, FP8, or BF16, GPU-selected) | vLLM (pip or docker) — multimodal (text + video) |
 | `agent-mcp-servers/transcript-mcp/` | `transcript_mcp_server` | 8200 | — | JSONL + FastMCP |
-| `agent-mcp-servers/video-mcp/` | `video_mcp_server` | 8210 | — | FastMCP → XR-Media-Hub |
+| `services/video-memory-service/` | `video_memory_service` | 8310 | — | Typed live/recorded video capability |
+| `agent-mcp-servers/video-mcp/` | `video_mcp_server` | 8210 | — | FastMCP → video-memory-service |
```
[the remark this answers is not in the document]
Comment (nvddr) [reply docs/source/components/ai-services.md]: Addressed in b2fde34: the component table now calls video-memory-service a typed recorded-video capability and identifies the MCP live IPC path.

**Comment by nvddr** [comment]
Disposition on the target-convergence review: native xr_video_memory intentionally keeps its recorded tools statically registered, matching the modular target; recording availability stays a runtime service/control concern, while Video MCP preserves its conditional legacy surface. The live/recorded ownership split is now aligned: LiveFrameSource owns current hub frames and video-memory-service owns persisted H.264/NVDEC history. The remaining schema-file/service-file layout differences are non-functional incremental structure, left in place to avoid churn before the later hub-client/package consolidation.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 13]
```diff
@@ -1,760 +1,198 @@
...
-        ready_file.touch()
+    run_dir = Path(config.get("out_dir") or "/tmp/xr_video_queries")
+    endpoint_task = asyncio.create_task(endpoint.run(), name="video-mcp-hub-ipc")
     try:
+        health = await client.get_health()
+        app = build_mcp(
+            client,
+            LiveFrameExporter(LiveFrameSource(endpoint), run_dir),
+            recording_enabled=health.recording_enabled,
+        ).http_app(path="/mcp")
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: Addressed in 11e0c57: startup now catches typed health-probe RPC errors, logs the condition, and builds the legacy live-only tool set. This preserves live-frame compatibility during a recorded-service outage; recorded tools appear after a later MCP restart once health succeeds. Added focused CPU coverage for the fallback.

**Comment by wenxind-nvidia** [comment]
**Minor — the recorded-discovery tool masks a transient outage as "empty," inconsistent with its sibling tools.**

`agent-mcp-servers/video-mcp/video_mcp_server/__main__.py::list_recorded_participants` catches `RPCError` and returns `[]` (with a `logger.warning`). Its siblings `get_video_stats` / `query_video` / `get_frame_from_time` instead surface the failure as `{"error": ...}`, and the native NAT `list_recorded_participants` lets the error propagate. So if the recorded service was up at startup (tools registered) but goes transiently down mid-session, an agent asking "who has recorded video?" gets an empty list indistinguishable from "nobody recorded" — and may tell the user there's no history when the service is merely unreachable.

Consider returning an error-as-data shape here too (or `{"participants": [], "unavailable": true}`), so discovery degrades the same way the other recorded tools do. (The startup `_recording_enabled → live-only` path is fine — it drops the recorded tools entirely and logs.)

**Review by wenxind-nvidia — APPROVED** [review]

File: services/video-memory-service/video_memory_service/store.py [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+                return self._check(directory)
+        return None
+
+    @staticmethod
+    def _matches(directory: Path, participant_id: str) -> bool:
+        identity = directory / ".identity"
+        if identity.exists():
+            return identity.read_text(encoding="utf-8") == participant_id
+        return directory.name == participant_id
+
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/store.py]: Addressed in de83758: participant directories and their .identity sidecars are now checked through the recordings-root trust boundary before any directory check or read. The same guard now covers metadata sidecars; regression coverage verifies an escaping directory symlink returns path_escape.

File: services/video-memory-service/video_memory_service/service.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            decoded_frames=len(frames),
+            target_us=target_us,
+        )
+        width = int(metadata.get("width", frames[index].shape[1]))
+        height = int(metadata.get("height", frames[index].shape[0] * 2 // 3))
+        rgb = await asyncio.to_thread(nv12_to_rgb, frames[index], width, height)
+        path = self._out_dir / (
+            f"{safe_name(request.participant_id)}_ago{request.second_ago}_{target_us}.png"
+        )
+        await asyncio.to_thread(save_png, rgb, path)
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/service.py]: Addressed in de83758: NV12 conversion and PNG writing now map failures to the stable frame_export_error RPC code. CPU coverage forces a conversion failure and asserts that contract.

File: services/video-memory-service/video_memory_service/store.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        for directory in sorted(self.root.iterdir()):
+            if not directory.is_dir():
+                continue
+            identity = directory / ".identity"
+            participants.append(
+                identity.read_text(encoding="utf-8") if identity.exists() else directory.name
+            )
+        return participants
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/store.py]: Addressed in de83758: participants() now validates each directory and .identity sidecar before reading it. The regression test also verifies an escaping identity symlink returns the stable path_escape error.

File: services/video-memory-service/README.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+```yaml
+endpoint: tcp://0.0.0.0:8310
+recordings_dir: /dev/shm/xr-ai/recordings
+out_dir: /tmp/xr_video_queries
+gpu_id: 0
+```
+
+`recordings_dir` must match XR Media Hub's `video_recording.out_dir`. Omit it
+only when running the service for a compatibility health check; all recorded
+operations then return `recording_disabled`.
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/README.md]: Addressed in 4b0efb1: the service README now distinguishes the empty list returned by list_recorded_participants when recording is disabled from the recording_disabled errors returned by the remaining recorded operations.

File: docs/source/components/mcp-servers.md [hunk 18]
```diff
@@ -236,23 +240,32 @@ enabled (a chunk store on disk):
...
+subscription:
+
+```yaml
 host:           0.0.0.0
 port:           8210
-gpu_id:         0
+service_endpoint: tcp://127.0.0.1:8310
+hub_pub:        ipc:///tmp/xr_hub_pub
+hub_push:       ipc:///tmp/xr_hub_in
+out_dir:        /tmp/xr_video_queries
```
[the remark this answers is not in the document]
Comment (nvddr) [reply docs/source/components/mcp-servers.md]: Addressed in 4b0efb1: the inline documentation now states that Video MCP out_dir applies only to live PNG exports; video-memory-service owns recorded query and frame output paths. The reference and sample YAML comments match.

**Comment by nvddr** [comment]
Disposition on the recorded-discovery outage: addressed in 4b0efb1. Video MCP now returns the normal participant list on success and an error object on a transient recorded-service failure, so it no longer represents an outage as empty history. This intentionally changes only the transitional MCP adapter result shape; native video functions continue to propagate typed RPC errors.

File: services/video-memory-service/video_memory_service/store.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+    def chunks(self, participant_id: str) -> list[tuple[Path, dict]]:
+        directory = self._participant_dir(participant_id)
+        if directory is None:
+            return []
+        paths = sorted(
+            (self._check(path) for path in directory.glob("*.264")),
+            key=lambda path: int(path.stem),
+        )
+        return [(path, self._metadata(path)) for path in paths]
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/store.py]: Addressed in 5e9af99: chunk discovery now skips non-timestamp .264 artifacts with a warning, so valid history remains queryable. A CPU regression test verifies the malformed file is ignored.

File: agent-sdk/xr_ai_agent/_live_frames.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+        self._timeout_s = timeout_s
+        self._latest: dict[tuple[str, str], FrameSignal] = {}
+        self._events: dict[str, asyncio.Event] = {}
+        endpoint.on_frame(self._on_frame)
+
+    async def _on_frame(self, signal: FrameSignal) -> None:
+        self._latest[(signal.participant_id, signal.track_id)] = signal
+        event = self._events.get(signal.participant_id)
+        if event is not None:
+            event.set()
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-sdk/xr_ai_agent/_live_frames.py:43]: Addressed in 30e305d: LiveFrameSource now releases cached signals on participant departure. The follow-up in 628410a keeps pending waiter events alive until their requests complete, so a departure/reconnect does not strand them.

File: tests/test_video_memory_functions.py [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+    await endpoint.send_participant(
+        ParticipantEvent(participant_id="departed-user", joined=False, pts_us=now_us)
+    )
+
+    assert source._latest == {}
+    assert source._events == {}
+
+    waiter.cancel()
+    with contextlib.suppress(asyncio.CancelledError):
+        await waiter
```
[the remark this answers is not in the document]
Comment (nvddr) [reply tests/test_video_memory_functions.py]: Addressed in 628410a: the regression now verifies that a fresh frame after reconnection resumes the pending get, and asserts its result instead of cancelling the task.

File: agent-sdk/xr_ai_agent/_live_frames.py [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        return frame
+
+    def release(self, participant_id: str) -> None:
+        """Drop cached signal and waiter state for a disconnected participant."""
+        self._latest = {
+            key: signal
+            for key, signal in self._latest.items()
+            if key[0] != participant_id
+        }
+        self._events.pop(participant_id, None)
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-sdk/xr_ai_agent/_live_frames.py]: Addressed in 628410a: each pending get owns a wait event, release wakes all pending events without dropping them, and each event is removed after its request completes. The regression covers a departure followed by a fresh reconnect signal.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 23]
```diff
@@ -1,760 +1,210 @@
...
-        # Anchored chunk lookup (recording path).
-        target_us = anchor_us - second_ago * 1_000_000
-        found = store.find_chunk_at(participant_id, target_us)
-        if found is None:
-            return {"error": f"No recorded video for {participant_id!r}. Recording may be disabled."}
-        chunk_path, meta = found
+            try:
+                return await live_frames.get_latest(participant_id)
+            except (FrameUnavailable, ValueError) as error:
+                return _error(error)
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: Addressed in fe548df: the shared live-frame helper now returns error-as-data for FrameUnavailable, ValueError, and PNG-export OSError failures. A CPU test covers this get_frame_from_time live path.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 24]
```diff
@@ -1,760 +1,210 @@
...
-    return mcp
+    if not recording_enabled:
 
+        @mcp.tool()
+        async def get_latest_frame(participant_id: str) -> dict:
+            """Return the current live frame. Deprecated; use get_frame_from_time."""
+            try:
+                return await live_frames.get_latest(participant_id)
+            except (FrameUnavailable, ValueError) as error:
+                return _error(error)
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: Addressed in fe548df: get_latest_frame now uses the same structured live-frame helper, including PNG-export OSError handling. The focused CPU test verifies both legacy live tools return an error object.

File: agent-mcp-servers/video-mcp/video_mcp_server/__main__.py [hunk 25]
```diff
@@ -1,760 +1,210 @@
...
+    parser.add_argument("--config", type=Path, default=None)
+    parser.add_argument("--ready-file", type=Path, default=None)
+    args, _ = parser.parse_known_args()
+
     setup_logging("video-mcp")
-    asyncio.run(_serve(cfg, ready_file=ns.ready_file))
+    config_path = args.config or _DEFAULT_CONFIG
+    if not config_path.exists():
+        sys.exit(f"video-mcp: config file not found: {config_path}")
+    config = yaml.safe_load(config_path.read_text()) or {}
```
[the remark this answers is not in the document]
Comment (nvddr) [reply agent-mcp-servers/video-mcp/video_mcp_server/__main__.py]: Addressed in fe548df: Video MCP now falls back to built-in defaults when its implicit source-tree YAML is absent from an installed wheel. A missing explicitly supplied --config still exits, and the CPU test covers both cases.

File: services/video-memory-service/video_memory_service/__main__.py [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument("--config", type=Path, default=None)
+    parser.add_argument("--ready-file", type=Path, default=None)
+    args, _ = parser.parse_known_args()
+
+    setup_logging("video-memory-service")
+    config_path = args.config or _DEFAULT_CONFIG
+    if not config_path.exists():
+        sys.exit(f"video-memory-service: config file not found: {config_path}")
+    config = yaml.safe_load(config_path.read_text()) or {}
```
[the remark this answers is not in the document]
Comment (nvddr) [reply services/video-memory-service/video_memory_service/__main__.py]: Addressed in fe548df: video-memory-service now uses built-in defaults when its implicit reference YAML is absent from an installed wheel, while preserving the failure for a missing explicit --config. The CPU test covers both entry points.

File: tests/test_video_memory_functions.py [hunk 27]
```diff
@@ -0,0 +1,447 @@
...
+import asyncio
+import contextlib
+import json
+import time
+import uuid
+from pathlib import Path
+from types import SimpleNamespace
+
+import pytest
+import video_mcp_server.__main__ as video_mcp_main
```
[the remark this answers is not in the document]
Comment (nvddr) [reply tests/test_video_memory_functions.py]: Addressed in 309fd40: the test now imports the Video MCP entry-point module once and references its helpers through that namespace.
